### PR TITLE
Bug 1878758: openstack UPI: Optional floating IPs

### PR DESCRIPTION
**Pull request by pierreprinetti**
Add the option to disable the use of floating IPs

/assign adduarte iamemilio

### CODE REVIEW

**Comment by iamemilio** [comment]
This seems reasonable

**Comment by pierreprinetti** [comment]
/retest
oh cmon Prow, we are entirely able to mess up by ourselves without your help

**Comment by iamemilio** [comment]
sorry about the review delay, but lgtm

**Comment by pierreprinetti** [comment]
Note for self:
Change os_use_fip to like bootstrap_fip. The other two can probably be set conditionally if the ip is not empty?

File: upi/openstack/inventory.yaml [hunk 1]
```diff
@@ -32,3 +28,16 @@ all:
...
       os_compute_nodes_number: 3
+
+      # If set to 'true', the 'bootstrap' playbook will create a floating IP
+      # attached to the bootstrap node, to enable the Installer to collect a
+      # 'must-gather' log bundle, useful for debugging a failed installation.
+      os_bootstrap_fip: true
+
+      # OpenShift API floating IP address. Will be attached if this value is
+      # non-empty
+      os_api_fip: '203.0.113.23'
```
Comment (iamemilio) [inline upi/openstack/inventory.yaml:39]: Do you think it is better to leave this as "", or to have a fake FIP in there? Does it get overwritten anyway and not matter?

File: upi/openstack/inventory.yaml [hunk 2]
```diff
@@ -32,3 +28,16 @@ all:
       # Number of provisioned Compute nodes.
       # 3 is the minimum number for a fully-functional cluster.
       os_compute_nodes_number: 3
+
+      # If set to 'true', the 'bootstrap' playbook will create a floating IP
+      # attached to the bootstrap node, to enable the Installer to collect a
+      # 'must-gather' log bundle, useful for debugging a failed installation.
+      os_bootstrap_fip: true
```
Comment (adduarte) [inline upi/openstack/inventory.yaml:35]: this is an extra variable. can't we ust "when os_api_fip is defined condition"?  in the playbooks

File: upi/openstack/network.yaml [hunk 3]
```diff
@@ -145,9 +145,11 @@
...
     command:
       cmd: "openstack floating ip set --port {{ os_port_api }} {{ os_api_fip }}"
+    when: os_api_fip
 
   # NOTE: openstack ansible module doesn't allow attaching Floating IPs to
   # ports, let's use the CLI instead
   - name: 'Attach the Ingress floating IP to Ingress port'
     command:
       cmd: "openstack floating ip set --port {{ os_port_ingress }} {{ os_ingress_fip }}"
+    when: os_ingress_fip
```
Comment (adduarte) [inline upi/openstack/network.yaml:155]: again here
"when: os_ingress_fip"

File: upi/openstack/network.yaml [hunk 4]
```diff
@@ -145,9 +145,11 @@
   - name: 'Attach the API floating IP to API port'
     command:
       cmd: "openstack floating ip set --port {{ os_port_api }} {{ os_api_fip }}"
+    when: os_api_fip
```
Comment (adduarte) [inline upi/openstack/network.yaml:148]: I think you want this to be 
"when: os_api_fip is defined"

File: upi/openstack/inventory.yaml (hunk 1 above)
Comment (pierreprinetti) [reply upi/openstack/inventory.yaml:39]: I find it useful that it matches the [documentation]. Is it confusing? I am worried about defaulting to FIPless, but you may have a better idea

[documentation]: https://github.com/openshift/installer/blob/master/docs/user/openstack/install_upi.md#api-and-ingress-floating-ip-addresses

**Comment by adduarte** [comment]
to align with IPI you should based it on weather or not external network and lbFloatingIP is defined, and also not create a router if no external network.

File: upi/openstack/inventory.yaml (hunk 1 above)
Comment (iamemilio) [reply upi/openstack/inventory.yaml:37]: I think matching the docs makes sense

File: upi/openstack/bootstrap.yaml [hunk 5]
```diff
@@ -40,3 +40,4 @@
       state: present
       network: "{{ os_external_network }}"
       server: "{{ os_bootstrap_server_name }}"
+    when: os_api_fip is defined and os_api_fip|length>0
```
Comment (mandre) [inline upi/openstack/bootstrap.yaml:43]: It's better to check if `os_external_network` was set, as it's a prerequisite for creating the FIP:
```suggestion
    when: os_external_network is defined and os_external_network|length>0
```

Comment (pierreprinetti) [reply upi/openstack/bootstrap.yaml:43]: What if the user sets the external network, but not the FIP? I found it interesting to use FIPs if passed, and not use them if not passed.

Maybe I could clarify in `inventory.yaml` that you should set an external network if you set the FIPs?

Comment (mandre) [reply upi/openstack/bootstrap.yaml:43]: > What if the user sets the external network, but not the FIP?

I'd argue this has nothing to do with the bootstrap node. And yes, this is likely going to break, unless they also modify the scripts to use an existing network configured with a router to an external network, and the FIP belong to the CIDR of the external subnet. But that's beyond the point.

For IPI, we do not allow setting a FIP without also setting an external network. Or at least that was the intent, https://github.com/openshift/installer/pull/3822#discussion_r454903882

Comment (pierreprinetti) [reply upi/openstack/bootstrap.yaml:43]: makes perfect sense :+1: thanks

File: upi/openstack/inventory.yaml [hunk 6]
```diff
@@ -32,3 +27,15 @@ all:
       # Number of provisioned Compute nodes.
       # 3 is the minimum number for a fully-functional cluster.
       os_compute_nodes_number: 3
+
+      # The public network, if available.
+      os_external_network: 'external'
+
+      # OpenShift API floating IP address. If this value is non-empty, the
+      # corresponding floating IP will be attached. Additionally if non-empty,
```
Comment (mandre) [inline upi/openstack/inventory.yaml:35]: This comment needs to be updated to remove the mention of bootstrap node FIP

Comment (pierreprinetti) [reply upi/openstack/inventory.yaml]: Good catch. While I was there, I have added some verbosity to the documentation.

**Comment by pierreprinetti** [comment]
>Timeout waiting for the server to come up

/test e2e-openstack-upi

**Comment by pierreprinetti** [comment]
hmm
/test e2e-openstack-upi

**Comment by pierreprinetti** [comment]
Depends on #4004

**Comment by pierreprinetti** [comment]
/cc @luis5tb 

Luis can you please check for Kuryr compatibility?

**Comment by pierreprinetti** [comment]
Needs https://github.com/openshift/release/pull/10944

**Comment by pierreprinetti** [comment]
Needs https://github.com/openshift/release/pull/11785

**Comment by pierreprinetti** [comment]
/retitle Bug 1878758: openstack UPI: Optional floating IPs

**Comment by pierreprinetti** [comment]
/hold cancel
CI passed!

**Review by mandre — APPROVED** [review]
/lgtm

**Pull request merged**